### PR TITLE
Merging PO(T) files

### PR DESCRIPTION
**Pull request by whatyouhide**
So this is the *first* and **very basic** version of merging PO(T) files. I'd like you @josevalim to take a look at this before continuing to work on it since I think your vision will be very useful here :).

Things I'd like to point your attention to:

* I kind of *had* to create a `Gettext.PO.Translations` module in order to keep all the translation-related functions in the same place (we normally would have put such functions in the `Gettext.PO.Translation` module but we have `PluralTranslation` as well :|). The name may be confusing and I'm still not sure about it, but without such a module those kind of functions would have ended up in `Gettext.PO` (as private functions, so no tests for them). This way they're nicely grouped in the same module.
* There are some grey areas when merging PO files, e.g., what do we do with headers? Do we want to keep the translations as much ordered as possible (I think so, VCS would be hell otherwise)?

Note that the code in this PR doesn't make a distinction between merging PO files, POT files or both. We'll want to do that but maybe we should call the process of merging a new POT file with old PO files "updating" instead of merging (e.g., POEdit calls it that way).

Looking forward to feedback :heart:

### CODE REVIEW

File: lib/gettext/po.ex [hunk 1]
```diff
@@ -234,4 +235,46 @@ defmodule Gettext.PO do
...
+    translations =
+      discard_autogenerated(unique_old) ++
+      merge_translations(common_old, common_new) ++
+      unique_new
+
+    %PO{translations: translations, headers: merge_headers(old.headers, new.headers)}
+  end
+
+  defp merge_translations(old, new) do
+    for {old_t, new_t} <- Enum.zip(Enum.sort(old), Enum.sort(new)) do
```
Comment (josevalim) [inline lib/gettext/po.ex]: I am afraid this Enum.sort/1 is going to be accidental and could change in the future when we change the struct. Is there a field we could choose for the sorting explicitly?

File: lib/gettext/po.ex [hunk 2]
```diff
@@ -234,4 +235,46 @@ defmodule Gettext.PO do
...
+  def merge(old, new) do
+    {common_old, unique_old} = Enum.partition old.translations, fn(t) ->
+      Enum.find new.translations, &Translations.same?(t, &1)
+    end
+
+    {common_new, unique_new} = Enum.partition new.translations, fn(t) ->
+      Enum.find old.translations, &Translations.same?(t, &1)
+    end
+
+    translations =
```
Comment (josevalim) [inline lib/gettext/po.ex]: You mentioned you want to keep the order as much as possible and this commit will move something a bit. Maybe the cleanest way to keep the order is to skip the first partition and use a flat map:

    Enum.flat_map old.translations, fn t ->
      cond do
        same = Enum.find new.translations, &Translations.same?(t, &1) ->
          [merge_translation(t, same)]
        autogenerated?(t) ->
          []
        true ->
          [t]
      end
    end
    
This way we also get rid of the sorting issues i mentioned below.

**Comment by josevalim** [comment]
This looks great! I have added some comments regarding the sorting but everything else looks perfect. :heart:

File: lib/gettext/po.ex (hunk 2 above)
Comment (whatyouhide) [inline lib/gettext/po.ex]: Neat!

**Comment by whatyouhide** [comment]
So, I'll work on this tonight but I wanted to write down some thoughts first. This PR is a little messy for what concerns the difference between PO vs POT files. `Gettext.PO.merge/2` as implemented now simply doesn't care about its arguments being PO or POT files. However, as we discussed in #17, merging POT files is slightly different than merging a POT file into a PO file.

I think we should tackle this; we could create a function to specifically merge POT files (which is very similar to the current `merge/2`, except it ignores `msgstr`s since they have to be empty in a POT file), say `merge_pot/2`. After that, we can create a function `merge_pot_into_po/2` which behaves slightly different, i.e., it removes all obsolete translations (the POT is the canonical reference). These two functions could share a bunch of the helper functions that `merge/2` currently uses, so there shouldn't be any plain repetition.

As an alternative to this, we can add a `:template?` (or `:pot?`) field to the `Gettext.PO` struct; the value of this field is determined when the struct is created, either by explicitly setting it or by checking that all the `msgstr`s are empty. This way we can keep only one `merge/2` function and make it polymorphic (since we can tell PO from POT files).

Wdyt @josevalim?

(PS we should give a thought to merging headers too, but that's a story for another day)

File: lib/gettext/po.ex (hunk 2 above)
Comment (whatyouhide) [inline lib/gettext/po.ex]: @josevalim implemented your suggestion almost verbatim, works like a charm!

**Comment by josevalim** [comment]
:shipit:

**Comment by josevalim** [comment]
> I think we should tackle this; we could create a function to specifically merge POT files (which is very similar to the current merge/2, except it ignores msgstrs since they have to be empty in a POT file), say merge_pot/2.

I don't think we need this. The msgstrs are going to be empty anyway... which means we don't need to do anything! :)

Btw, we should probably look at gettext for the operation names. Check this out: http://www.gnu.org/software/gettext/manual/gettext.html#index-msgmerge. Our code is almost the same except our code keeps the non autogenerated and theirs do not. So I think the only thing we need to do is to add an option like `:keep_non_autogenerated` (or a better name) and call it a day. We probably don't need separate functions even.

Also, we need to discuss what we are going to do with duplicated entries in the same .po file. What if we have the same msgid twice? It will probably emit a warning when compiling but the warnings won't be clear. We can probably keep a HashDict when we load the translations storing all `msgid => line`, if there is a duplicate, we could store it in a list which we return as result of parse, something like: `{:dup_msgid, [line1, lin2, msgid]}`. Then the entity calling parse can decide if it is going to raise, print errors, etc. This is the approach used by Erlang's compiler.

**Comment by whatyouhide** [comment]
d20947e adds support for the `:keep_manual_translations` option in `Gettext.PO.merge/2`; this option defaults to `true` (current behaviour) but can be set to `false` if we want to discard all obsolete translations (that are not in the reference - PO or POT - file).

This option will likely let us avoid making separate functions for PO/POT files as we can reproduce the behaviour of a newer POT file being merged into an older PO file by setting `:keep_manual_translations` to `false`, meaning the POT file becomes the absolute reference for the merge.

File: lib/gettext/po.ex [hunk 3]
```diff
@@ -234,4 +235,57 @@ defmodule Gettext.PO do
...
   defp escape_char(?\r),  do: ~S(\r)
   defp escape_char(char), do: <<char>>
+
+  @doc """
+  Merges two `Gettext.PO` structs, giving precedence to the second one.
+
+  This function is used to merge two `Gettext.PO` structs into one. `new` has
+  precedence over `old`, meaning that:
+
+    * all headers from `old` are discarded (only the ones from `new` are kept)
```
Comment (josevalim) [inline lib/gettext/po.ex]: I think we want to keep po headers when merging pot into po and also all comments. I think we already merge comments, right?

Comment (whatyouhide) [inline lib/gettext/po.ex]: Absolutely. This is how I dealt with headers when I wrote this, quoted directly from the PR 

> (PS we should give a thought to merging headers too, but that's a story for another day)

:) Comments are correctly merged: no translator `#[space]` comment is discarded (they're merged and put one after the other), reference comments from the old file are ditched.

I'm playing with `msgmerge` now and indeed it looks like only headers from PO files are kept (headers in POT files are ignored). Nice and easy :)

**Comment by whatyouhide** [comment]
I fixed the handling of headers in `Gettext.PO.merge/2`, I'm going to finally merge this :).

**Pull request merged**